server/graphql/v2/query/collection/ExpensesCollectionQuery.ts

```diff
@@ -374,13 +375,35 @@ export const ExpensesCollectionQueryResolver = async (
   }
 
   const order = [[args.orderBy.field, args.orderBy.direction]] as OrderItem[];
+
   const { offset, limit } = args;
-  const result = await models.Expense.findAndCountAll({ include, where, order, offset, limit });
+
+  let nodes;
+
+  const fetchNodes = () => {
+    if (!nodes) {
+      // We don't await on purpose, we set and return the Promise
+      nodes = Expense.findAll({ include, where, order, offset, limit });
+    }
+    return nodes;
+  };
+
+  const fetchTotalCount = async () => {
+    await sleep(1); // Give time for fetchNodes to be triggered
+    if (nodes) {
+      const result = await nodes;
+      if (!offset && (!limit || result.length < limit)) {
+        return result.length;
+      }
+    }
+    return Expense.count({ include, where });
+  };
+
   return {
-    nodes: result.rows,
-    totalCount: result.count,
+    nodes: fetchNodes,
+    totalCount: fetchTotalCount,
```

Comment:
Can we maybe have a shared helper for that? It sounds like we'd want the same approach in different places, and it would help to contain the complexity of this resolver. Something like:

```ts
const { nodes, totalCount } = await paginateResults(Expense, { include, where, order, offset, limit });
```

Reply:
I removed this part.